### ISSUE TITLE
EntityManager find with pessimistic lock does not check for transaction when cached

### ISSUE THREAD

**Issue by madwizard-thomas** [doctrine/doctrine2#7068]
When a pessimistic lock mode is passed to EntityManager::find it normally throws a TransactionRequiredException if no transaction is currently active.

However if the entity is currently in the identity map, it will refresh the entity from the database and return it but not check if a transaction is running. The refresh query has 'FOR UPDATE' (in case of pessimistic write) but this has no effect since there is no transaction.

To reproduce:
```php
// $em is EntityManager, Order entity can be any entity
$em->find(Order::class, 1); // Load order id 1 in entity map
$em->find(Order::class, 1, LockMode::PESSIMISTIC_WRITE); // does not throw exception
$em->find(Order::class, 2, LockMode::PESSIMISTIC_WRITE); // throws TransactionRequiredException 
```
Tested with doctrine 2.5.14 but latest code seems to have the same issue.

**Comment by Ocramius** [comment]
Is this a MySQL specific problem?

**Comment by madwizard-thomas** [comment]
I don't think so, but have only tested it with MySQL. It seems to be caused by the early return statement [here](https://github.com/doctrine/doctrine2/blob/v2.5.14/lib/Doctrine/ORM/EntityManager.php#L435) before the transaction check [here](https://github.com/doctrine/doctrine2/blob/v2.5.14/lib/Doctrine/ORM/EntityManager.php#L455)

**Comment by BreiteSeite** [comment]
Thanks @madwizard-thomas for the minimal code to reproduce. Helped a lot.

@Ocramius i added this as a test for the test-suite and also provided a fix in #7075

**Comment by Ocramius** [comment]
Handled in #7075 and #7291
